space/space_test.go

```diff
@@ -6,248 +6,260 @@ import (
-	return test.assertErrorType(errors.DataConflictError{})
-}
-
-func (test *repoBBTest) assertErrorType(e error) func(p *space.Space, e2 error) {
-	return func(p *space.Space, err error) {
-		assert.Nil(test.T(), p)
-		assert.IsType(test.T(), e, err, "error was %v", err)
-	}
-}
-
-func (test *repoBBTest) requireErrorType(e error) func(p *space.Space, err error) {
-	return func(p *space.Space, err error) {
-		assert.Nil(test.T(), p)
-		require.IsType(test.T(), e, err)
-	}
-}
-
-func (test *repoBBTest) create(name string) func() (*space.Space, error) {
-	newSpace := space.Space{
-		Name:    name,
-		OwnerId: test.testIdentity.ID,
-	}
-	return func() (*space.Space, error) { return test.repo.Create(context.Background(), &newSpace) }
-}
-
-func (test *repoBBTest) save(p space.Space) func() (*space.Space, error) {
-	return func() (*space.Space, error) { return test.repo.Save(context.Background(), &p) }
-}
-
-func (test *repoBBTest) load(id uuid.UUID) func() (*space.Space, error) {
-	return func() (*space.Space, error) { return test.repo.Load(context.Background(), id) }
+		// when updating this space
+		sp, err := s.repo.Save(s.Ctx, &p)
+		// then
+		require.NotNil(t, err)
+		require.IsType(t, errors.NotFoundError{}, err, "error was %v", err)
+		require.Nil(t, sp)
+	})
 }
 
-func (test *repoBBTest) loadByUserIdAndName(userId uuid.UUID, spaceName string) func() (*space.Space, error) {
-	return func() (*space.Space, error) {
-		return test.repo.LoadByOwnerAndName(context.Background(), &userId, &spaceName)
-	}
+func (s *repoBBTest) TestDelete() {
+	s.T().Run("ok", func(t *testing.T) {
+		// given a space
+		fxt := tf.NewTestFixture(t, s.DB, tf.Spaces(1))
+		id := fxt.Spaces[0].ID
+		// double check that we can load this space
+		sp, err := s.repo.Load(s.Ctx, id)
+		require.Nil(t, err)
+		require.NotNil(t, sp)
+		// when
+		err = s.repo.Delete(s.Ctx, id)
+		// then
+		require.Nil(t, err)
+		// double check that we can no longer load the space
+		sp, err = s.repo.Load(s.Ctx, id)
+		require.NotNil(t, err)
+		require.IsType(t, errors.NotFoundError{}, err, "error was %v", err)
+		require.Nil(t, sp)
+	})
+	s.T().Run("not found - not existing space ID", func(t *testing.T) {
+		// given a not existing space ID
+		nonExistingSpaceID := uuid.NewV4()
+		// when
+		err := s.repo.Delete(s.Ctx, nonExistingSpaceID)
+		// then
+		require.NotNil(t, err)
+		require.IsType(t, errors.NotFoundError{}, err, "error was %v", err)
+	})
+	s.T().Run("not found - nil space ID", func(t *testing.T) {
+		// given a not existing space ID
+		nilSpaceID := uuid.Nil
+		// when
+		err := s.repo.Delete(s.Ctx, nilSpaceID)
+		// then
+		require.NotNil(t, err)
+		require.IsType(t, errors.NotFoundError{}, err, "error was %v", err)
+	})
 }
 
-func (test *repoBBTest) delete(id uuid.UUID) func() (*space.Space, error) {
-	return func() (*space.Space, error) { return nil, test.repo.Delete(context.Background(), id) }
+func (s *repoBBTest) TestList() {
+	s.T().Run("ok", func(t *testing.T) {
```

Comment:
Does this test deal with `pagination` too? And should we create more than 1 spaces because we are testing list!

Reply:
If you look closely at what I refactored, I did not invent new tests but just removed all the hassle with the helper functions and made the test more explicit and grouped them with tests and sub-tests. Still your question is valid.